File: Leetcode/0198-HouseRobber/go/solution1.go
func rob(nums []int) int {
	n := len(nums)
	if n == 1 {
		return nums[0]
	}
	// Define dp[i] as max gain of robbing houses [i, n)
	dp := make([]int, n+1) // for convenience, dp[n] = 0
	// Base case
	dp[n-1] = nums[n-1] // Only 1 house available
	for i := n - 2; i >= 0; i-- {
		dp[i] = max(nums[i]+dp[i+2], dp[i+1])
	}
	return dp[0]
}

File: Leetcode/0198-HouseRobber/java/Solution1.java
class Solution {
    public int rob(int[] nums) {
        // Define dp[i] as the maximum amount of money you can rob froms nums[0, i)
        int[] dp = new int[nums.length + 1];
        // Base case: only 1 house
        dp[1] = nums[0];
        for (int i = 2; i <= nums.length; i++) {
            // Two possibilities:
            // 1. Rob nums[i-1] and don't rob nums[i-2]
            // 2. Rob nums[i-2] and don't rob nums[i-1]
            dp[i] = Math.max(dp[i - 1], dp[i - 2] + nums[i - 1]);
        }
        return dp[nums.length];
    }
}